File: Chapter1/c++ programs/unique_path.cpp
// 计数型动态规划
#include<iostream>
#include<string>
#include<vector>
using namespace std;

class Solution {
public:
    /**
     * @param m: positive integer (1 <= m <= 100)
     * @param n: positive integer (1 <= n <= 100)
     * @return: An integer
     */
    int uniquePaths(int m, int n) {
        // write your code here
        int f[m][n];
        // 计算顺序是逐行逐列的
        for(int i = 0; i < m; i++){
            for (int j=0; j < n; j++){
                // 初始化条件(边界情况，这里已经隐含了f[0][0]=1的边界情况了)
                if (i == 0 || j == 0){
                    f[i][j] = 1;
                }
                else{
                    // 上边 + 左边
                    f[i][j] = f[i-1][j] + f[i][j-1];
                }
            }
        }
        return f[m-1][n-1];
    }
};

int main(){
    Solution s;
    int m= 3;
    int n = 3;
    int result = s.uniquePaths(m, n);
    cout << "result = " << result << endl;
}

File: Chapter1/python programs/unique_path.py
# 方法1：常规方式
class Solution:
    def uniquePaths(self, m, n):
        # write your code here
        if m == 0 or n == 0:
            return -1
        # 初始化一个多维数组的正确方式：
        f = [[0 for j in range(n)] for i in range(m)]
        # 下面这种方式不对，因为*m只是把对象的引用复制了m次, 因此f[0][0]修改时，f[i][0]也会更改
        # f = [[0]*n]*m
        for i in range(m):
            for j in range(n):
                if i == 0 or j == 0:
                    f[i][j] = 1
                else:
                    f[i][j] = f[i-1][j] + f[i][j-1]
        return f[m-1][n-1]

# 方法2：滚动数组
"""
class Solution:
    def uniquePaths(self, m, n):
        # write your code here
        if m == 0 or n == 0:
            return -1
        # 初始化一个多维数组的正确方式：
        f = [[0 for j in range(n)] for i in range(2)]
        old, now = 0, 0
        # 下面这种方式不对，因为*m只是把对象的引用复制了m次, 因此f[0][0]修改时，f[i][0]也会更改
        # f = [[0]*n]*m
        for i in range(m):
            old = now
            now = 1 - now 
            for j in range(n):
                if i == 0 or j == 0:
                    f[now][j] = 1
                else:
                    f[now][j] = f[old][j] + f[now][j-1]
        return f[now][n-1]
"""

if __name__ == '__main__':
    s = Solution()
    m, n = 3, 3
    result = s.uniquePaths(m, n)
    print("result = ", result)